docs/fobnail-data-format.md

```diff
@@ -0,0 +1,87 @@
+# Fobnail data format
+
+Fobnail uses CBOR (with a few exceptions) for transmitting structured data both
+from Fobnail to Attester and from Attester to Fobnail. We use subset of
+[RFC7049](https://datatracker.ietf.org/doc/html/rfc7049) (no tagging and no
+extensions).
+
+## EK certificate
+
+EK certificate is transferred as a raw DER. We don't add any more data so there
+is no need to wrap it into CBOR.
+
+## AIK
+
+AIK is transferred in TPM format (TPM2B_PUBLIC structure). Re-encoding this into
+CBOR would strip essential data. Fobnail needs unmodified TPM2B_PUBLIC to
+successfully perform the "make credential" operation which is needed for
+credential activation.
+
+## Metadata
+
+Metadata is sent in CBOR-encoded format (also described
+[here](https://github.com/fobnail/docs/blob/main/dev-notes/metadata.md) in more
+detail). Example metadata looks like this (for readability, presented as JSON)
+
+```json
+{
+    // metadata version, used to detect backward-incompatible changes
+    "version": 1,
+    "manufacturer": "Gigabyte",
+    "model": "Gigabyte A520 AORUS ELITE",
+    "mac": [0xea, 0x96, 0x91, 0x87, 0x26, 0x8d],
+    "sn": "S21N559B431",
+}
+```
+
+Metadata is signed by Attester using AIK before sending (see
+[Data signing](#data-signing) section below).
+
+## RIM
+
+RIM holds PCR registers (from TPM) which are used for attestation. It doesn't
+contain information about the Attester since these are part of metadata. RIM
+may contain multiple PCR banks (depending of hash algorithm: SHA-1, SHA-256,
+others). Each PCR bank contains bitmap of present PCRs (`pcrs` field) and array
+of PCR itself. The MSB of bitmap corresponds to pcr31 and LSB to pcr0.
+
+```json
+{
+    // PCR update counter
+    "update_ctr": 0,
+    // First PCR bank
+    "sha1": {
+        "pcrs": 0xffffffff,
+        "pcr": [
+            [ 0x00, 0x00, 0x00, 0x00, ... ],
+            ...
+        ]
+    },
+    // Another PCR bank
+    "sha2": {
+        "pcrs": 0x000000ff,
+        "pcr": [
+            [ 0x00, 0x00, 0x00, 0x00, ... ],
+            ...
+        ]
+    }
+}
+```
+
+RIM is signed by Attester using AIK before sending (see
+[Data signing](#data-signing) section below).
+
+## Data signing
+
+Signature is generated by taking an already CBOR-encoded object, a nonce
+(generated by Fobnail and sent as part of request) and signing hash of them
+both. The data is wrapped into another CBOR object:
+
+```json
+{
+    // The data we signed (nested CBOR)
+    "data": [0x42, ...],
```

Comment:
That doesn't tell us where exactly nonce is added (prepended vs. appended), nor whether it is added to CBOR or as plain bytes before/after CBOR.

Reply:
Fixed